wfl/fit/ref_error.py

```diff
@@ -1,213 +1,240 @@
+                data = at.arrays
+
+            # grab data
+            ref_quant = data.get(ref_property_prefix + prop_use)
+            calc_quant = data.get(calc_property_prefix + prop_use)
+            if ref_quant is None or calc_quant is None:
+                # skip if data is missing
+                warnings.warn(f"Missing property ref {ref_property_prefix + prop_use} is None {ref_quant is None} "
+                              f"or calc {calc_property_prefix + prop_use} is None {calc_quant is None}")
+                continue
+
+            if virial_from_stress:
+                # ref quant was actually stress, automatically convert
+                ref_quant *= -at.get_volume()
+                calc_quant *= -at.get_volume()
+
+            # make everything into an appropriately shaped and normalized array
+            ref_quant = _reshape_normalize(ref_quant, prop, at, per_atom)
+            calc_quant = _reshape_normalize(calc_quant, prop, at, per_atom)
+
+            if prop in config_properties or not by_species:
+                # If quantities are not being split up by species, make a
+                # group that includes all atoms, so loop below that separates out
+                # things by Z will lump them all together.
+                atom_split_indices = np.asarray(range(len(ref_quant)))
+                atom_split_groups = [(atom_split_indices, "")]
+            else: # atom_properties
+                # Make separate groups for each atomic number Z, so their errors
+                # can be tabulated separately.
+                Zs = at.numbers
+                atom_split_indices = np.asarray(Zs)
+                atom_split_groups = [(Z, f"_{Z}") for Z in sorted(set(Zs))]
+
+            # Loop over groups that errors should be split up by within each configuration.
+            # Only atomic number Z implemented so far (see above).
+            for atom_split_index_val, atom_split_index_label in atom_split_groups:
+                # use only subset of quantities that are relevant to this subset of atoms,
+                # normally either all atoms or the ones with one particular atomic number
+                ref_quant = ref_quant[atom_split_indices == atom_split_index_val]
+                calc_quant = calc_quant[atom_split_indices == atom_split_index_val]
+
+                if per_component:
+                    # if per component, flatten all vector component so each is counted separately
+                    ref_quant = ref_quant.reshape((-1, 1))
+                    calc_quant = calc_quant.reshape((-1, 1))
+
+                diff = calc_quant - ref_quant
+
+                if len(diff.shape) != 2:
+                    raise RuntimeError(f"Should never have diff.shape={diff.shape} with dim != 2 (prop {prop + atom_split_index_label})")
+                # compute norm along vector components
+                diff = np.linalg.norm(diff, axis=1)
```

Comment:
if the property (e.g. forces) are not per-component, shouldn't we take the norm of reference and predicted/calculated property too?

Reply:
You mean the ones that are returned in the form of the difference and parity?  That's going to be a bit messy.  For the differences, I think I was intending to return in more or less the original form, although it's possible that I actually ended up flattening all the remaining array indices. For the parity it's not obvious to me.  I guess we should take the norm, although I'm not sure that a parity plot of force norms is all that interesting - arguably the parity plot should always be per-component.

Reply:
> You mean the ones that are returned in the form of the difference and parity?

yes

So I'm plotting parity plot and an error scatter plot beneath. This came up because I had individual components for reference (x axis) and magnitude of differences for y axis (3 times fewer entries). Not sure if reshaping and normalising components just for plot is safe enough? Alternatively, could always return components, but return both errors - per component and of magnitudes.

Reply:
If you're using the parity information for a scatter plot, rather than a parity plot, then something like the norm of the force makes sense.  How about we _do_ apply the norm to the parity but only if it's not per-component? You can always request the property regular and per-component, and use the first for a scatterplot and the second for a parity plot.

Reply:
> How about we do apply the norm to the parity but only if it's not per-component?  

Sounds good.

Reply:
OK, I'll code that up.

Reply:
I think I have added that in one of my commits - maybe you had something else in mind

Reply:
No, I hadn't realized you had done it already.  I'll review your changes and discuss (or just fix) anything I think needs it.

Reply:
I'm happy with the normalization fix, but I'm making small changes to the plotting script.